Containerd image and snapshot backend integration

dmcgowan opened the pull request:
Opening this up as a draft as there are many things left to do before this is mergeable.

Highlights
- Image and reference store deprecated for containerd backend
- ImageID can reference manifests, manifest lists, or image configs (allowing better support for multi-arch images)
- Support for multiple unpacked storage backends (images point to their backend)
- Multi-arch Image support

Major things left to do
- Migration of existing image store data
- Build code
- Artifact creation on push and commit
- Snapshot support as an option for images to point to
- Many, many TODOs

Implements #38043

Does not include, both snapshotters and graphdriver interfaces will be supported in this PR
* Graph drivers scoped down (no more mounting/diffing)
* Graph drivers supported as containerd snapshotter plugins

Closes #26146
Closes #38043 
Closes https://github.com/moby/moby/pull/33355
Closes https://github.com/docker/cli/pull/122

dmcgowan committed:
Use containerd client to pull images

Signed-off-by: Derek McGowan <derek@mcgstyle.net>

dmcgowan committed:
Use containerd to list images

Signed-off-by: Derek McGowan <derek@mcgstyle.net>

dmcgowan committed:
Use containerd to create tags

Signed-off-by: Derek McGowan <derek@mcgstyle.net>

dmcgowan committed:
Add image cache

Use cache for delete and listing images

Signed-off-by: Derek McGowan <derek@mcgstyle.net>

dmcgowan committed:
Add image tagging by id and reference

Signed-off-by: Derek McGowan <derek@mcgstyle.net>

dmcgowan committed:
Add image deletion

Fix conflict and update removal to use cache and not stores.
Layer removal is still missing.

Signed-off-by: Derek McGowan <derek@mcgstyle.net>

dmcgowan committed:
Add support for layers

Unpack and delete layers

Signed-off-by: Derek McGowan <derek@mcgstyle.net>

dmcgowan committed:
Add support for inspect

Signed-off-by: Derek McGowan <derek@mcgstyle.net>

dmcgowan committed:
Update image history to use containerd

Signed-off-by: Derek McGowan <derek@mcgstyle.net>

dmcgowan committed:
Add support for docker run

Signed-off-by: Derek McGowan <derek@mcgstyle.net>

dmcgowan committed:
Update image service functions to use containerd

Updates info functions to query the containerd image cache.
Updates related call stack to pass down context.

Signed-off-by: Derek McGowan <derek@mcgstyle.net>

Commit:
display docker pull progress for containerd integration

Signed-off-by: Anda Xu <axu@knox.edu>

dmcgowan committed:
Add support for commit

Fix dangling image management to not rely on reference parsing

Signed-off-by: Derek McGowan <derek@mcgstyle.net>

dmcgowan committed:
Add parent label to committed images

Fix layer references being taken on startup

Signed-off-by: Derek McGowan <derek@mcgstyle.net>

dmcgowan committed:
Update events handler to use containerd

Signed-off-by: Derek McGowan <derek@mcgstyle.net>

dmcgowan committed:
Add push support through containerd

Uses containerd client to push.
Temporarily resolves current platform on push to match current behavior.

Signed-off-by: Derek McGowan <derek@mcgstyle.net>

dmcgowan committed:
Update images to use manifest as image ID

Lookup images without using caching

Signed-off-by: Derek McGowan <derek@mcgstyle.net>

AkihiroSuda commented:
Is this planned to be merged before Docker v19.03?

Commit:
unpack layers simultaneously

Signed-off-by: Anda Xu <axu@knox.edu>

dmcgowan commented:
> Is this planned to be merged before Docker v19.03?

That is desirable but unrealistic. I want to complete as many or all of the major things by that time frame, but until all the functionality is complete it is too hard to pin down an exact date for "stable enough to merge". 

It is very important we get the design parts mentioned in "highlights" correct so we aren't in a position where we need to correct storage code later (adding more support permutations and instability). Please comment on the design and help wherever possible will certainly help stabilize this change.

dmcgowan committed:
Add layer label to content

Adds equivalent label as the snapshotter gc label

Signed-off-by: Derek McGowan <derek@mcgstyle.net>

AkihiroSuda commented on daemon/images/image.go:
```diff
@@ -1,11 +1,39 @@
+const (
+	// LabelImageID refers to the image ID used by Docker
+	// Deprecate this to support multi-arch images
+	LabelImageID = "docker.io/image.id"
```
nit: `mobyproject.org`? `docker.com`?

thaJeztah replied:
If these can conflict with user-provided labels; reverse-dns notation was originally marked as reserved; https://docs.docker.com/config/labels-custom-metadata/#key-format-recommendations

> The `com.docker.*`, `io.docker.*`, and `org.dockerproject.*` namespaces are reserved by Docker for internal use.

dmcgowan committed:
Resolve runtime image

Signed-off-by: Derek McGowan <derek@mcgstyle.net>

Commit:
unblock layer extracting to a separate thread

Signed-off-by: Anda Xu <axu@knox.edu>

dmcgowan committed:
Move layer creation to daemon package

Signed-off-by: Derek McGowan <derek@mcgstyle.net>

dmcgowan committed:
Support for multiple layer stores

Make layer stores multiplatform and allow configuring
multiple ordered by preference.

Signed-off-by: Derek McGowan <derek@mcgstyle.net>

cpuguy83 commented:
Pushed a rebase.

dmcgowan committed:
Update pull to use error groups

Support for multiple unpacks

Signed-off-by: Derek McGowan <derek@mcgstyle.net>

dmcgowan committed:
Update rmi to use containerd gc

Signed-off-by: Derek McGowan <derek@mcgstyle.net>

dmcgowan committed:
Lookup images by digest regexp

Filter pull artifacts by platform while pulling
all manifests and configs.

Signed-off-by: Derek McGowan <derek@mcgstyle.net>

dmcgowan committed:
Use content label for inspecting layer store

Signed-off-by: Derek McGowan <derek@mcgstyle.net>

dmcgowan committed:
Update pull to add canonical tag

Update tag to avoid deprecate cache

Signed-off-by: Derek McGowan <derek@mcgstyle.net>

Commit:
use containerd client to create image during build

Signed-off-by: Anda Xu <axu@knox.edu>

dmcgowan committed:
Add test framework which sets up containerd

Add initial test for listing images

Signed-off-by: Derek McGowan <derek@mcgstyle.net>

GordonTheTurtle commented:
<!-- AUTOMATED:POULE:DCO-EXPLANATION -->
Please sign your commits following these rules:
https://github.com/moby/moby/blob/master/CONTRIBUTING.md#sign-your-work
The easiest way to do this is to amend the last commit:
~~~console
$ git clone -b "containerd-integration" git@github.com:dmcgowan/docker.git somewhere
$ cd somewhere
$ git rebase -i HEAD~842359063784
editor opens
change each 'pick' to 'edit'
save the file and quit
$ git commit --amend -s --no-edit
$ git rebase --continue # and repeat the amend for each commit
$ git push -f
~~~

Amending updates the existing PR. You **DO NOT** need to open a new one.

dmcgowan committed:
Add delete image unit tests

Signed-off-by: Derek McGowan <derek@mcgstyle.net>

GordonTheTurtle commented:
<!-- AUTOMATED:POULE:DCO-EXPLANATION -->
Please sign your commits following these rules:
https://github.com/moby/moby/blob/master/CONTRIBUTING.md#sign-your-work
The easiest way to do this is to amend the last commit:
~~~console
$ git clone -b "containerd-integration" git@github.com:dmcgowan/docker.git somewhere
$ cd somewhere
$ git rebase -i HEAD~842359062744
editor opens
change each 'pick' to 'edit'
save the file and quit
$ git commit --amend -s --no-edit
$ git rebase --continue # and repeat the amend for each commit
$ git push -f
~~~

Amending updates the existing PR. You **DO NOT** need to open a new one.

dmcgowan committed:
Fix pull lease cleanup to use new context

Prevents lease getting left around after canceled context

Signed-off-by: Derek McGowan <derek@mcgstyle.net>

dmcgowan committed:
Add more rmi tests

Add test for removing parent and child.
Add checks for layer removal.

Signed-off-by: Derek McGowan <derek@mcgstyle.net>

dmcgowan committed:
Add os platform matcher

Update backends to match only on platform

Signed-off-by: Derek McGowan <derek@mcgstyle.net>

dmcgowan committed:
Update commit to migrate layers

Signed-off-by: Derek McGowan <derek@mcgstyle.net>

dmcgowan committed:
Update history to not rely on deprecated cache

Signed-off-by: Derek McGowan <derek@mcgstyle.net>

dmcgowan committed:
Show digest tags from dangling images

Signed-off-by: Derek McGowan <derek@mcgstyle.net>

dmcgowan committed:
Add size to docker images

Signed-off-by: Derek McGowan <derek@mcgstyle.net>

dmcgowan committed:
Improve platform matching for containers and pull

By default match a single architecture platform for
pull. Use container platform for resolving layerstore.
Use container driver name when available to resolve
layerstore.

Signed-off-by: Derek McGowan <derek@mcgstyle.net>

dmcgowan committed:
Remove deprecated cache

Use layer cache when necessary or containerd client directly

Signed-off-by: Derek McGowan <derek@mcgstyle.net>

dmcgowan committed:
Fix docker images filters

Support dangling only and reference lookups

Signed-off-by: Derek McGowan <derek@mcgstyle.net>

dmcgowan committed:
Optimize image list to skip size on skipped images

Signed-off-by: Derek McGowan <derek@mcgstyle.net>

kolyshkin committed:
image: restore(): prevent panic

This is to avoid nil pointer dereference.

Signed-off-by: Kir Kolyshkin <kolyshkin@gmail.com>

kolyshkin committed:
Preliminary image store migration support

This migrates image store to containerd.

To test:
 # sudo DOCKER_MIGRATE_IMAGE_STORE=da ./bundles/dynbinary-daemon/dockerd-dev

To re-test:
 # ctr -n moby image ls -q | xargs ctr -n moby image rm
 # ctr -n moby content ls -q | xargs ctr -n moby content rm

TODO:
 - find a better place for this code
 - maybe some tests?

Signed-off-by: Kir Kolyshkin <kolyshkin@gmail.com>

dmcgowan committed:
Fix image lookup and listing

Ensure ambiguous reference is not mistakenly returned.
Cleanup TODOs in listing and resolve image

Signed-off-by: Derek McGowan <derek@mcgstyle.net>

dmcgowan committed:
Fix unit tests and lint

Signed-off-by: Derek McGowan <derek@mcgstyle.net>

dmcgowan committed:
Fix CI linting

Signed-off-by: Derek McGowan <derek@mcgstyle.net>

dmcgowan committed:
Revert image package changes

The entire package is now deprecated, revert temporary
changes.

Signed-off-by: Derek McGowan <derek@mcgstyle.net>

Commit:
add already exists image check during pull

Signed-off-by: Anda Xu <axu@knox.edu>

dmcgowan committed:
Support updating existing image after pull

Fix build

Signed-off-by: Derek McGowan <derek@mcgstyle.net>

dmcgowan committed:
Fix swagger nonsense

Signed-off-by: Derek McGowan <derek@mcgstyle.net>

dmcgowan committed:
Pass test debug flag from environment

Signed-off-by: Derek McGowan <derek@mcgstyle.net>

dmcgowan committed:
Add load implementation using containerd import

Signed-off-by: Derek McGowan <derek@mcgstyle.net>

dmcgowan committed:
Update make emptyfs to use supported load format

Signed-off-by: Derek McGowan <derek@mcgstyle.net>

dmcgowan committed:
Update output to match integration tests

Signed-off-by: Derek McGowan <derek@mcgstyle.net>

Commit:
clean up code and add comments to image_builder

Signed-off-by: Anda Xu <axu@knox.edu>

Commit:
resolve runtime image properly for both builder and end user
fix already exists error during image tagging

Signed-off-by: Anda Xu <axu@knox.edu>

dmcgowan committed:
Move migration to separate function

Remove image store instantiation from image service

Signed-off-by: Derek McGowan <derek@mcgstyle.net>

dmcgowan committed:
Fix validation

Signed-off-by: Derek McGowan <derek@mcgstyle.net>

dmcgowan committed:
Update migration to use layer stores

Move logic to image service and merge with cache loading.

Signed-off-by: Derek McGowan <derek@mcgstyle.net>

GordonTheTurtle commented:
Looks like we would not be able to merge this PR because of merge conflicts. Please rebase, fix conflicts, and force push to your branch.

dmcgowan committed:
Ignore not exists errors on migration

Signed-off-by: Derek McGowan <derek@mcgstyle.net>

dmcgowan committed:
Update dockerfile builder

Replace use of image id with descriptor

Signed-off-by: Derek McGowan <derek@mcgstyle.net>

dmcgowan committed:
Add default tag on resolve image name

Signed-off-by: Derek McGowan <derek@mcgstyle.net>

dmcgowan committed:
Use containerd exporter

Update load to use full image name annotation

Signed-off-by: Derek McGowan <derek@mcgstyle.net>

dmcgowan committed:
Fix integration tests

Prevent panics on unimplemented code and remove
unnecessary lookups on commit.

Signed-off-by: Derek McGowan <derek@mcgstyle.net>

dmcgowan committed:
Remove unnecessary layer store OS check

Signed-off-by: Derek McGowan <derek@mcgstyle.net>

andrewrynhard commented:
Is this planned for Docker 19.03? I have a use case where I would like to `--output=type=docker,dest=example.tar`, but this is not supported for multi-arch containers.

thaJeztah commented:
No, this is not for 19.03; much work to do yet before this can be merged

dt-rush commented:
Is there any way we might separate out the multi-arch image support and merge that sooner? This seems it's a long time running since February and has slowed down a fair bit. Lack of multi-arch image support is preventing `docker buildx --load` from loading manifest list "images" into the docker daemon.

dmcgowan committed:
Handle tagless names when matching short identifiers

Signed-off-by: Derek McGowan <derek@mcgstyle.net>

carlosedp commented:
Any news about this? It's pretty important for local development with buildx avoiding problems like https://github.com/docker/buildx/issues/59.
Thanks!

CameronMcWilliam commented:
Also interested to see if this has made any progress recently! :)

t0rr3sp3dr0 commented:
@dmcgowan do you plan to continue developing this?

milahu commented:
oh, i see. im still surprised, cos `docker load < oci.tar` would be trivial to implement (no need for containerd), since the only difference between oci and docker are the file locations, e.g. `layerhash.tar` vs `blobs/sha256/layerhash` (please tell me im terribly wrong ^^). imho oci is better, cos `sha256sum` is much faster than `tarhash`

tianon commented:
I don't think you're wrong, but this is not the right place to discuss it 😅

(A new PR would be very interesting 👀)

RoyHP commented:
This would be great to see, especially now with Apple Silicon creating a barrier to entry for people trying to run amd64 containers in the cloud but arm64 locally on their machines.

dmcgowan commented:
I'm going to close this one since I don't want anyone to get the impression I am still working on it. The branches live on in my fork if anyone wants to play around with it. Reaching full behavior parity and just keeping this branch updated is non-trivial.

pkit commented:
@tianon 
> I don't think you're wrong, but this is not the right place to discuss it sweat_smile

I'm not sure how come it's not the right place to discuss it? 
This issue is a direct descendant of the original issue #25779 (opened in 2016!) that wanted exactly that: `docker load < oci.tar`
But okay, let's reopen #25779 then.
